Assembly-CSharp/ModLoader.cs

```diff
@@ -488,36 +510,56 @@ IEnumerator PreloadScene(string s)
                         modScenePreloadedObjects[objName] = obj;
                     }
                 }
-
-                // Update loading progress
-                progress++;
-
-                updateLoadingBarProgress();
-                yield return USceneManager.UnloadSceneAsync(scene);
-                updateLoadingBarProgress();
             }
 
-            List<IEnumerator> batch = new();
-            int maxKeys = toPreload.Keys.Count;
-
-            foreach (string sceneName in toPreload.Keys)
+            void CleanupPreloadOperation(string sceneName)
             {
-                int batchCount = Math.Min(ModHooks.GlobalSettings.PreloadBatchSize, maxKeys);
-
-                batch.Add(PreloadScene(sceneName));
-
-                if (batch.Count < batchCount)
-                    continue;
+                Logger.APILogger.LogFine($"Unloading scene \"{sceneName}\"");
+                var unloadOperation = USceneManager.UnloadSceneAsync(sceneName);
+                unloadOperation.completed += (op) =>
+                {
+                    sceneProgress[sceneName] = 1.0f;
+                    preloadOperationQueue.Remove(unloadOperation);
+                };
+                preloadOperationQueue.Add(unloadOperation);
+            }
 
-                Coroutine[] coros = batch.Select(nb.StartCoroutine).ToArray();
+            void StartPreloadOperation(string sceneName)
+            {
+                Logger.APILogger.LogFine($"Loading scene \"{sceneName}\"");
+                sceneProgress[sceneName] = 0.0f;
+                var preloadOperation = USceneManager.LoadSceneAsync(sceneName, LoadSceneMode.Additive);
+                preloadOperation.priority = scenePriority[sceneName];
+                preloadOperation.completed += (op) =>
+                {
+                    sceneProgress[sceneName] = 0.5f;
+                    preloadOperationQueue.Remove(preloadOperation);
+                    GetPreloadObjectsOperation(sceneName);
+                    CleanupPreloadOperation(sceneName);
+                };
+                preloadOperationQueue.Add(preloadOperation);
+            }
 
-                foreach (var coro in coros)
-                    yield return coro;
+            int i = 0;
+            while (sceneProgress.Values.Average() < 1.0)
+            {
+                while (
+                    preloadOperationQueue.Count < ModHooks.GlobalSettings.PreloadBatchSize &&
+                    i < sceneNames.Count &&
+                    sceneProgress.Values.Average() < 1.0
+                )
+                {
+                    StartPreloadOperation(sceneNames[i++]);
+                }
+                // live update
+                // divided by 2 because each preload is made of 2 parts, loading and unloading, each giving 0.0 - 1.0 as progress
+                updateLoadingBarProgress(sceneProgress.Values.Sum() + (preloadOperationQueue.Select(x => x.progress).Sum() / 2.0f));
```

Comment:
Calculating the sum of the progress values twice a frame is non-ideal, though it might be fine just for the sake of being more readable.